[CLI] Add paging support for output

tobwen opened the pull request:
### Summary
Implements pager functionality for shell query results:

- Supports three modes: auto, on and off (default)
- Uses system PAGER environment variable when available
- Falls back to 'less -S' on Unix and 'more' on Windows
- Allows setting custom pager command
- Adds new .pager command with documentation
- Automatically resets output after paged content

Paging makes it easier to view large query results by leveraging system pager utilities rather than scrolling through all output.

### Example
```
.mode csv
.pager '/usr/bin/local/pspg --csv'
```

### Discussion
https://github.com/duckdb/duckdb/discussions/13427

tobwen committed:
Add paging support for query output

Implements pager functionality for shell query results:

- Supports three modes: auto, on and off (default)
- Uses system PAGER environment variable when available
- Falls back to 'less -S' on Unix and 'more' on Windows
- Allows setting custom pager command
- Adds new .pager command with documentation
- Automatically resets output after paged content

Paging makes it easier to view large query results by leveraging
system pager utilities rather than scrolling through all output.

Mytherin commented:
Thanks for the PR!

Trying this out on MacOS - this doesn't really seem to work well with `.mode duckbox` because of the ANSI escape codes used for rendering colors, e.g.:

```sql
.pager on
from range(100);
```

Results in:

```
ESC[90m┌ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m>
ESC[0mESC[90m│ESC[0mESC[90m   ESC[0mrangeESC[90m    ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m   ESC[0mint64ESC[90m    ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m├ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mESC[90m─ESC[0mE>
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m0ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m1ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m2ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m3ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m4ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m5ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m6ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m7ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m8ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m          ESC[0m9ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m10ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m11ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m12ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m13ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m14ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m15ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m16ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m17ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m18ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m19ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0mESC[90m ·ESC[0mESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0mESC[90m ·ESC[0mESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0mESC[90m ·ESC[0mESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m80ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m81ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m82ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m83ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m84ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m85ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m86ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m87ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m88ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m89ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m90ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m91ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m92ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m93ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m94ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m95ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m96ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
ESC[0mESC[90m│ESC[0mESC[90m         ESC[0m97ESC[90m ESC[0mESC[90m│ESC[0mESC[90m
```

Maybe this is down to my pager not supporting this, and there are pagers that do. 

Maybe for paging we want to set the default settings to:

```
.maxrows -1
.maxwidth -1
.highlight_results off
```


In Postgres, what happens is that the pager is only opened when the query result is larger than a certain amount of rows. e.g:

```sql
postgres=# select * from generate_series(0, 10);
-- small result, no pager
postgres=# select * from generate_series(0, 1000);
-- pager
```

I think this is a desirable feature that we would want to have as well by default.

tobwen committed:
Remove AUTO pager mode and simplify pager configuration

tobwen committed:
Add required includes for pager error handling
Add SIGPIPE signal handling for pager operations

tobwen committed:
Enhance pager environment variable handling and default behavior

tobwen committed:
Update help documentation for simplified pager commands

tobwen committed:
Improve pager setup with better error handling and user feedback

tobwen committed:
Add pager status to configuration display

tobwen committed:
Add string trimming utility function

tobwen committed:
Refactor pager command handling with improved validation

tobwen committed:
Refine pager implementation

Improvements made:
- Move C++ standard library includes (algorithm, cctype, locale) to the top of the file with other includes for proper organization
- Add error handling for pclose() to detect and report pager failures
- Improve .pager status output to show both mode and effective command
- Update help documentation to clarify that .pager without args shows status
- Add comments to helper functions for better code documentation
- Improve output formatting consistency (capitalization, alignment)

These changes enhance code quality, maintainability, and user experience without altering the core pager functionality.

tobwen committed:
Add comprehensive test suite for CLI pager functionality

Implements 31 test cases covering all aspects of the pager feature:

Core Functionality Tests:
- Default pager status and configuration
- Pager on/off toggling
- Custom pager commands with arguments
- Environment variable handling (PAGER and DUCKDB_PAGER)
- DUCKDB_PAGER takes precedence over PAGER

Integration Tests:
- Pager with different output modes (CSV, duckbox, columnar)
- Pager with query output (small and large result sets)
- Pager with output redirection to files
- Pager with headers and null values
- Pager with multiple queries in sequence
- Pager with transactions

Edge Cases:
- Empty string disables pager
- Command trimming
- Invalid pager command handling
- Error messages not affected by pager
- Configuration persistence across queries
- Mode changes don't affect pager state

All tests follow DuckDB shell test patterns using the ShellTest framework and pytest. Tests verify both stdout and stderr output as appropriate.

tobwen committed:
Add pager command validation and documentation

- Implement isPagerAvailable() function to validate pager commands
- Check if pager executable exists and is accessible before use
- Display warning when pager command is not found or not executable
- Add PAGER_IMPLEMENTATION.md documenting pager functionality
- Support both absolute/relative paths and PATH lookups
- Cross-platform support for Windows and Unix-like systems

tobwen committed:
Fix pager validation: don't enable pager when command is not found or executable

When the pager isn't found or executable, it shouldn't be used. This change ensures that:
1. Pager mode is not set to 'on' when the command is invalid
2. Invalid pager commands are not saved to pager_command
3. initializePager() validates system pagers before setting them

Previously, the shell would accept any pager command and only fail at runtime when trying to execute it. Now, validation happens upfront using isPagerAvailable() to check if the command exists and is executable.

tobwen committed:
Refine pager validation: preserve state when invalid command is provided

Instead of turning pager OFF when an invalid command is provided, keep the current pager state unchanged. This provides better user experience by maintaining their pager configuration.

tobwen committed:
Add -pager <cmd> command-line parameter

- Adds -pager option to set pager command from command line
- Validates that pager command exists and is executable
- Displays warning if pager command is not found or not executable
- Uses existing isPagerAvailable() validation logic
- Supports commands with arguments (e.g., 'less -SR')
- Command-line argument overrides environment variables
- Added to help text in alphabetical order

tobwen committed:
Add pager validation with priority fallback and warnings

Implement comprehensive pager validation that emits warnings when pagers are not found or not executable. The system now tries all pagers in priority order until a valid one is found.

Priority order (highest to lowest):
1. -pager command-line argument
2. DUCKDB_PAGER environment variable
3. PAGER environment variable
4. Platform default (more on Windows, none on Unix)

Changes:
- Modified getSystemPager() to validate environment variables and emit warnings for invalid pagers
- Reordered functions so isPagerAvailable() comes before getSystemPager() to avoid compilation errors
- Updated initializePager() to accept command-line pager argument and implement priority-based fallback with warnings
- Moved initializePager() call to after command-line argument parsing so the -pager argument can be captured and validated
- Updated command-line argument parsing to capture -pager in first pass
- Removed duplicate -pager handling in second pass
- Used fprintf() instead of utf8_printf() for early warnings (before macro definition)

tobwen committed:
Improve pager command help documentation

Removes the detailed pager implementation documentation file as it's no longer needed in the main codebase.

Updates the help text for the pager command to be more concise and clearer:
- Aligns text spacing for better readability
- Simplifies descriptions of the on/off commands
- Makes the custom command explanation more straightforward
